### PR TITLE
Add GitHub Actions to automate updates and checks 

### LINKED ISSUE #203: Add a check to detect files with missing extensions
There've been a few cases of files missing their `.md` extension - enough that I think it's worth adding a check somewhere to detect this.

### PR DESCRIPTION
**Pull request by claremacrae**
- This is the initial implementation of #153
- See the top of that issue for likely remaining steps
- **This is initially a draft PR, for discussion with @argenos on 9th Jan 2022**

## Edited

- `.github/scripts/tests/test_make_mocs.py` no longer users a hard-coded specific diff tool when showing any failures in approval tests
- `.github/scripts/tests.py` now returns a non-zero exit code if the tests fail
- `.github/scripts/utils.py` updated to not write out any progress bars if being run inside GitHub Actions (by recognising the standard environment variable that is for this purpose)

## Added

- Add new script `.github/scripts/check_content.py` which implements the first check for any files with no file extension: this fixes #203
- Add 2 GitHub composite actions, to share consistent instructions between workflows:
    - `.github/actions/setup-python/action.yml`
        - Sets up Python - currently 3.8
    - `.github/actions/install-dependencies/action.yml`
        - Installs dependencies in the `requirements.txt` file
- Add 3 GitHub Actions workflows, to run automated processes:
    - `.github/workflows/check_content.yml`
        - Runs `check_content.py` on every push and pull request on all branches
    - `.github/workflows/python.yml`
        - Runs the python unit tests on every push and pull request on all branches
        - May later be extended to run extra checks, such as if we add typehints in future
    - `.github/workflows/update_hub.yml`
        - Adds and updates various bits of data in the Hub content
           - theme download counts
           - new plugins, themes and authors
           - moc files
       - Creates a pull request, and shows the link to the new PR
       - Will likely be triggered via a daily cron job - which means it will only run on the main branch

## Checklist

*Not applicable*

### CODE REVIEW

File: .github/scripts/check_content.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+from os import walk
+
+
+def check_content() -> int:
+    """
+    Walks through the filetree rooted at the current working directory.
+    For each file that it finds, it validates the file
+    """
+    error_count = 0
+    DIRECTORIES_TO_EXCLUDE = ['.git', '.github', '.idea', 'venv', 'DO NOT COMMIT']
```
Comment (dbarenholz) [inline .github/scripts/check_content.py:11]: Some people (me) call their virtual environment `.venv`. How should we deal with all possibilities of virtual environment names? Maybe a addition (or creation if not yet exists) to a note on "scripts for the hub" explaining to add your virtual environment directory to this list?

File: .github/scripts/check_content.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    Writes error messages to the console, and returns the number of errors found
+
+    :param relative_path: The path of the file, including the directory, relative to the vault's root
+    :param file: The name of the file, without any directory
+    :return: The number of errors found in the given file 
+    """
+    errors = 0
+
+    if '.' not in file:
+        print(f'Error:\n  This file has no extension: consider adding ".md" to its name:\n  {relative_path} ')
```
Comment (dbarenholz) [inline .github/scripts/check_content.py:36]: Perhaps it would be a good idea to decide on a universal "logging" method/syntax/manner. My own opinion is that 1 error = 1 line, in stead of 3.

File: .github/scripts/check_content.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        errors += 1
+
+    # Other checks may be added here in future
+    return errors
+
+
+def get_root_of_vault() -> str:
```
Comment (dbarenholz) [inline .github/scripts/check_content.py:43]: Missing documentation.
"""
Helper method that returns the root directory of the vault.
"""

File: .github/scripts/check_content.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    up1 = os.path.dirname(dir_path)
+    up2 = os.path.dirname(up1)
+    return up2
```
Comment (dbarenholz) [inline .github/scripts/check_content.py:47]: I would add a comment (either # ... or in docstring) that this only works for files in the script category. If ever in the future we have `.github/scripts/script_type/some_script.py`, this no longer works as expected.

File: .github/scripts/check_content.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def main() -> int:
```
Comment (dbarenholz) [inline .github/scripts/check_content.py:50]: Docstring?

File: .github/scripts/check_content.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    os.chdir(get_root_of_vault())
+    return check_content()
+
+
+if __name__ == "__main__":
```
Comment (dbarenholz) [inline .github/scripts/check_content.py:55]: Do we want to explicitly log behaviour when this file is imported in stead of run? If so, we can do
```python
if __name__ == "__main__":
    # stuff
else:
    print(f"The script {__file__} is not meant to be imported!")
```

File: .github/workflows/check_content.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+name: check_content
+
+on:
+  # Intentionally check on all branches, to detect any issues as early as possible
```
Comment (dbarenholz) [inline .github/workflows/check_content.yml:4]: Do we really want to check this on all branches, or only on PRs? What if we have "non-working code" that we need to push because others want to work on it?

File: .github/workflows/python.yml [hunk 8]
```diff
@@ -0,0 +1,22 @@
+name: python
+
+on:
```
Comment (dbarenholz) [inline .github/workflows/python.yml:3]: Same comment as `check_content.yaml`.

**Comment by dbarenholz** [comment]
I have one more comment to think about related to `requirements.txt`, though perhaps this should go elsewhere: How do we want to deal with people adding dependencies? Is it a 'free for all'? Do we want to control which dependencies are added?

File: .github/scripts/check_content.py (hunk 1 above)
Comment (claremacrae) [reply .github/scripts/check_content.py:11]: I'm not sure how to encode this feedback in the code, but might first thought was to add a comment before that code saying something like:

```py
# Note to future developers: if a development-related directory gets in the way here,
# feel free to add it to this list
```

On further thought, I think a better idea is to re-use the code in `MocFileAndDirectoryFilter` in `make_mocs.py` which has more sophisticated logic in it, like excluding everything - files and directories - that begins with a `.` - which would simplify the code here greatly.

File: .github/scripts/check_content.py (hunk 2 above)
Comment (claremacrae) [reply .github/scripts/check_content.py:36]: We sort of have a logging facility already, which is in `utils.py`.

However, this isn't logging, it's an ~~error~~ help  message that must never be suppressed.

> My own opinion is that 1 error = 1 line, in stead of 3.

Ah, do you mean the output should be on one line? It started like this, but was quite hard to parse, so I wrapped it up.

Yeah, I would have previously said the same, until I saw a tweet talking about the mindset of writing "help" messages rather than "error" messages... It completely changed how I phrased such things.

The output now looks like this:

image

The path can get quite long - and so it's done with the file path on its own line, for glancability.
Similarly, the layout tries make it easy to spot the suggestion to add `.md`.

File: .github/scripts/check_content.py (hunk 3 above)
Comment (claremacrae) [reply .github/scripts/check_content.py:43]: Thank you - added - and credited....

File: .github/scripts/check_content.py (hunk 4 above)
Comment (claremacrae) [reply .github/scripts/check_content.py:47]: I must admit that I had assumed that all that mattered was the relative location from the source of this function... so it wouldn't matter if we called it from other levels.

To test this assumption, I just added a test for this function, which runs on my machine and CI, so it looks to me like we are OK on this... 

Or have I missed something? :-)

The test:
https://github.com/claremacrae/obsidian-hub/blob/22c2db7ca776b74ce4c360bc9dacbc81c893e924/.github/scripts/tests/test_check_content.py#L6-L9

The test run:
https://github.com/claremacrae/obsidian-hub/runs/4750104035?check_suite_focus=true#step:5:13

File: .github/scripts/check_content.py (hunk 5 above)
Comment (claremacrae) [reply .github/scripts/check_content.py:50]: I see what you mean...

On reflection, I think at the moment I'd rather leave the code as the documentation - as this seems like the sort of thing that might become out of date if it's duplicated.

But that prompted me to make these two renaming, so the function names themselves are at least self-documenting:

- `main()` -> `check_content_of_vault()
- `check_content()` -> `check_content_of_working_directory()`

Thanks for the prod!

File: .github/scripts/check_content.py (hunk 6 above)
Comment (claremacrae) [reply .github/scripts/check_content.py:55]: It's fine for this to be imported - and it was useful to be able to import it in a test, to write a test for part of its behaviour...

File: .github/workflows/check_content.yml (hunk 7 above)
Comment (claremacrae) [reply .github/workflows/check_content.yml:4]: In this case, "non-working code" would be markdown content that is broken...

I'm curious to understand what about having these checks turned on would prevent someone from pushing code in a non-working state?

File: .github/workflows/python.yml (hunk 8 above)
Comment (claremacrae) [reply .github/workflows/python.yml:3]: Yes, and same reply :-)

> I'm curious to understand what about having these checks turned on would prevent someone from pushing code in a non-working state?

Especially as this one is about of Python code is broken, and it's absolutely fine for someone to push code in an intermediate state where the tests are broken - but all the more reason for the tests to be running, so that it is known that the tests are failing...

**Comment by claremacrae** [comment]
> I have one more comment to think about related to `requirements.txt`, though perhaps this should go elsewhere: How do we want to deal with people adding dependencies? Is it a 'free for all'? Do we want to control which dependencies are added?

TL;DR: Let's not worry about that until or unless it becomes a real, practical problem.

It depends on what you mean by 'free for all'? 

I'm a relatively new addition to this team. But I will say that there wasn't a `requirements.txt` before I added one, so I don't think it's a big deal.

FWIW I would assume that future maintainers would generally only new dependencies if they were valuable to the project, and here wasn't something built-in that was good enough.

We do have code review to review that, and we will soon have CI to detect if some non-working dependency was added - or the perhaps more likely error that someone added a dependency but forgot to add it to the requirements...

So for now, I feel it's fine - and don't want to potentially scare other members of the team by adding more docs/rules for people to remember or maintain.

File: .github/scripts/check_content.py (hunk 2 above)
Comment (claremacrae) [reply .github/scripts/check_content.py:36]: Bear in mind that if this was on one line, it would wrap horribly in the GitHub Actions log.
And if there were several files missing their suffix, the single-line messages would be even harder to parse.

File: .github/scripts/check_content.py (hunk 5 above)
Comment (claremacrae) [reply .github/scripts/check_content.py]: Maybe I'm misunderstanding the kind of docs you meant?

File: .github/workflows/python.yml (hunk 8 above)
Comment (claremacrae) [reply .github/workflows/python.yml:3]: General experience about dealing with failing tests is to try and discover them as soon as possible... The longer it's left before discovering them, the more commits there are that will be possible candidates, and the more time-consuming it will be to fix.

If people commit little-and-often, and push frequently, I would expect the failing test to be discovered very soon after the breakage was created, which would make it much easier to spot the point the problem happened - and then either fix it, or ask for help fixing it....

**Comment by claremacrae** [comment]
@dbarenholz Thanks for really helpful code-review comments!

File: .github/scripts/check_content.py (hunk 1 above)
Comment (dbarenholz) [reply .github/scripts/check_content.py:11]: Whichever you decide to choose I'll agree with - just wanted to raise awareness. (feel free to resolve)

File: .github/scripts/check_content.py (hunk 5 above)
Comment (dbarenholz) [reply .github/scripts/check_content.py]: Yes: a docstring is this.
```python
def something():
   """
   This is a docstring explaining what the method does.
   """
```

File: .github/workflows/check_content.yml (hunk 7 above)
Comment (dbarenholz) [reply .github/workflows/check_content.yml:4]: Ah. Brainfart on my side that this isn't much of a coding project. Non-issue!

File: .github/scripts/check_content.py (hunk 5 above)
Comment (claremacrae) [reply .github/scripts/check_content.py]: > Maybe I'm misunderstanding the kind of docs you meant?

Hah - I was totally ambiguous - I was meaning what sort of information was worth documenting, that wasn't already clear from the code itself.

But I will write some general overall statement about the types of checks being run, like that it's checking the vault info/contents, rather than the infrastructure.

File: .github/scripts/check_content.py (hunk 1 above)
Comment (claremacrae) [reply .github/scripts/check_content.py]: > Some people (me) call their virtual environment `.venv`. How should we deal with all possibilities of virtual environment names? Maybe a addition (or creation if not yet exists) to a note on "scripts for the hub" explaining to add your virtual environment directory to this list?

@dbarenholz This was really helpful... It prompted me to make these two simplifications:

Re-use MocFileAndDirectoryFilter to filter directories
https://github.com/obsidian-community/obsidian-hub/pull/213/commits/a5c38734e607af8bf3f5af35c08c34b83d6264f1

Simplify the check for dot files in check_content.py
https://github.com/obsidian-community/obsidian-hub/pull/213/commits/0f4efb21a69eed88d5209a461255b2582ab74dea

File: .github/scripts/check_content.py (hunk 5 above)
Comment (claremacrae) [reply .github/scripts/check_content.py]: I've added a `--help` arg to `check_content.py`

```
python3 ./check_content.py --help
usage: check_content.py [-h]

Check for issues with the content of the Hub vault (such as errors in file names).

optional arguments:
  -h, --help  show this help message and exit
```

And added a comment at the top of the file:

```python
# This code is for validating the **Content** of the Hub, such as file names,
# as opposed to validating the Python code and other infrastructure.
```

Comment (claremacrae) [reply .github/scripts/check_content.py]: Thanks @dbarenholz - see above for the changes I made. Resolving conversation now.

**Review by argenos — APPROVED** [review]
Great work @claremacrae! Thank you!
